includes/REST_API/Products_Controller.php

```diff
@@ -0,0 +1,489 @@
+	 * @param Story_Post_Type $story_post_type Story_Post_Type instance.
+	 */
+	public function __construct( Settings $settings, Story_Post_Type $story_post_type ) {
+		$this->settings        = $settings;
+		$this->story_post_type = $story_post_type;
+
+		$this->namespace = 'web-stories/v1';
+		$this->rest_base = 'products';
+	}
+
+	/**
+	 * Get the list of service IDs required for this service to be registered.
+	 *
+	 * Needed because the story post type needs to be registered first.
+	 *
+	 * @since 1.20.0
+	 *
+	 * @return string[] List of required services.
+	 */
+	public static function get_requirements(): array {
+		return [ 'settings', 'story_post_type' ];
+	}
+
+	/**
+	 * Registers routes for links.
+	 *
+	 * @since 1.20.0
+	 *
+	 * @see register_rest_route()
+	 */
+	public function register_routes(): void {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			[
+				[
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => [ $this, 'get_items' ],
+					'permission_callback' => [ $this, 'permissions_check' ],
+					'args'                => $this->get_collection_params(),
+				],
+				'schema' => [ $this, 'get_public_item_schema' ],
+			]
+		);
+	}
+
+	/**
+	 * Checks if a given request has access to get and create items.
+	 *
+	 * @since 1.20.0
+	 *
+	 * @return true|WP_Error True if the request has read access, WP_Error object otherwise.
+	 */
+	public function permissions_check() {
+		if ( ! $this->story_post_type->has_cap( 'edit_posts' ) ) {
+			return new \WP_Error(
+				'rest_forbidden',
+				__( 'Sorry, you are not allowed to manage publisher logos.', 'web-stories' ),
+				[ 'status' => rest_authorization_required_code() ]
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Retrieves all products.
+	 *
+	 * @since 1.20.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_items( $request ) {
+		$results = [];
+
+		// TODO(#11154): Refactor to extract product query logic out of this controller.
+		$shopify_host         = $this->settings->get_setting( Settings::SETTING_NAME_SHOPIFY_HOST );
+		$shopify_access_token = $this->settings->get_setting( Settings::SETTING_NAME_SHOPIFY_ACCESS_TOKEN );
+
+		if ( ! empty( $shopify_host ) && ! empty( $shopify_access_token ) ) {
+			return $this->get_items_shopify( $request );
+		}
+
+		if ( function_exists( 'wc_get_products' ) ) {
+			return $this->get_items_woocommerce( $request );
+		}
+
+		return rest_ensure_response( $results );
+	}
+
+	/**
+	 * Retrieves all Shopify products.
+	 *
+	 * @SuppressWarnings(PHPMD.ExcessiveMethodLength)
+	 *
+	 * @since 1.20.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	protected function get_items_shopify( $request ) {
```

Comment:
I kind of feel like this code should maybe live in another class. This code is related to get shopify products and is not related to the endpoints. Having this as another class, should be massively useful for mocking and testing.